gson/src/main/java/com/google/gson/internal/bind/ReflectiveTypeAdapterFactory.java

```diff
@@ -226,34 +263,75 @@ private Map<String, BoundField> getBoundFields(Gson context, TypeToken<?> type,
 
   static abstract class BoundField {
     final String name;
+    /** Name of the underlying field */
+    final String componentName;
     final boolean serialized;
     final boolean deserialized;
 
-    protected BoundField(String name, boolean serialized, boolean deserialized) {
+    protected BoundField(String name, String componentName, boolean serialized, boolean deserialized) {
       this.name = name;
+      this.componentName = componentName;
       this.serialized = serialized;
       this.deserialized = deserialized;
     }
-    abstract void write(JsonWriter writer, Object value) throws IOException, IllegalAccessException;
-    abstract void read(JsonReader reader, Object value) throws IOException, IllegalAccessException;
+
+    abstract void write(JsonWriter writer, Object value) throws IOException, ReflectiveOperationException;
+
+    /** Read the value into an array of values, used to construct records */
+    abstract void readIntoArray(JsonReader reader, int index, Object[] values) throws IOException
+
+    /** Read the value from the reader, and set it on the corresponding field on value via reflection */
+    abstract void readIntoField(JsonReader reader, Object value) throws IOException, IllegalAccessException;
```

Comment:
I think here and in the override the `value` parameter should be renamed to `instance` (or something else). Personally I feel `value` is extremely misleading in the context of a field.
```suggestion
    /** Read the value from the reader, and set it on the corresponding field on instance via reflection */
    abstract void readIntoField(JsonReader reader, Object instance) throws IOException, IllegalAccessException;
```

The same might also apply to `write(JsonWriter writer, Object value)`.

(Though to be fair this code already existed before your PR.)

What do you think?

Reply:
Renamed to source and target, as then these match with each other on read and write.